Improve json conversion with tests and support for older Rails 3.x

cheremukhin23 opened the pull request:
- Add error handling when unable to parse json
- Add tests to ensure json_safe_and_pretty handles hashes and strings

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/shakacode/react_on_rails/787)
<!-- Reviewable:end -->

justin808 commented:
Looking great!

---

Reviewed 1 of 2 files at r1, 1 of 1 files at r2.
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---

*[app/helpers/react_on_rails_helper.rb, line 223 at r2](https://reviewable.io:443/reviews/shakacode/react_on_rails/787#-KgchU4ZQq1Xoe8SFvfI:-KgchU4ZQq1Xoe8SFvfJ:bn2krdq) ([raw file](https://github.com/shakacode/react_on_rails/blob/a3ef532d26039dc87d3032142a3171a6d26b2f3a/app/helpers/react_on_rails_helper.rb#L223)):*
> ```Ruby
>   end
> 
>   def json_safe_and_pretty(hash_or_string)
> ```

Can we throw here if the value passed is not a hash or a string? What else could the value be?

---

*[app/helpers/react_on_rails_helper.rb, line 231 at r2](https://reviewable.io:443/reviews/shakacode/react_on_rails/787#-KgcggrDdvDqTLt3uR8q:-KgcggrDdvDqTLt3uR8r:b-bw5u2q) ([raw file](https://github.com/shakacode/react_on_rails/blob/a3ef532d26039dc87d3032142a3171a6d26b2f3a/app/helpers/react_on_rails_helper.rb#L231)):*
> ```Ruby
>       escape_json(hash_value.to_json)
>     end
>   rescue JSON::ParserError => e
> ```

Can we move this catch block around the JSON.parse above?

we can do something like:

```ruby
  hash_value = if hash_or_string.is_a?(String) 
                          begin
                              JSON.parse(hash_or_string)
                           rescue JSON::ParserError => e
                               raise e, %(Cannot parse #{hash_or_string} \n Backtrace: #{e.backtrace.join("\n")})
                          else
                              hash_or_string
                          end  

```

---

*[app/helpers/react_on_rails_helper.rb, line 242 at r2](https://reviewable.io:443/reviews/shakacode/react_on_rails/787#-KgcgcN4lVDUvQARMTnr:-KgcgcN4lVDUvQARMTns:b75wqvp) ([raw file](https://github.com/shakacode/react_on_rails/blob/a3ef532d26039dc87d3032142a3171a6d26b2f3a/app/helpers/react_on_rails_helper.rb#L242)):*
> ```Ruby
>   end
> 
>   def old_json_escape(json)
> ```

Can you provide a comment with a link on to the github source of this?

---

*[spec/dummy/spec/helpers/react_on_rails_helper_spec.rb, line 40 at r2](https://reviewable.io:443/reviews/shakacode/react_on_rails/787#-KgchiN8u2axRSMV9z4x:-KgchiN8u2axRSMV9z4y:b-sl8qp0) ([raw file](https://github.com/shakacode/react_on_rails/blob/a3ef532d26039dc87d3032142a3171a6d26b2f3a/spec/dummy/spec/helpers/react_on_rails_helper_spec.rb#L40)):*
> ```Ruby
>       expect(escaped_json).to eq(hash_sanitized)
>     end
>   end
> ```

Can we add a couple tests that confirm the exceptions:

1. non-hash, non-string passed
2. non parseable string in development mode


Also, we're never testing the part where we check we're in development mode.

I'd try to mock/stub out the Rails.env for that test.

```ruby
# RSpec version
it 'tests my code' do
  allow(Rails).to receive(:env).and_return(ActiveSupport::StringInquirer.new('development'))
  # test code
end
```

[Some options from Google](https://www.google.com/search?q=stub+rails+env+for+test+to+production)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/shakacode/react_on_rails/787#-:-KgciraoQ9vwhFuV8sTf:b-cxhjm2)*
<!-- Sent from Reviewable.io -->

justin808 commented:
See note below. I might just release a temp patch on this.

---

Review status: all files reviewed at latest revision, 5 unresolved discussions.

---

*[app/helpers/react_on_rails_helper.rb, line 224 at r2](https://reviewable.io:443/reviews/shakacode/react_on_rails/787#-KggEw9tKZ5IIRJ7FLl0:-KggEw9tKZ5IIRJ7FLl1:bgrsuja) ([raw file](https://github.com/shakacode/react_on_rails/blob/a3ef532d26039dc87d3032142a3171a6d26b2f3a/app/helpers/react_on_rails_helper.rb#L224)):*
> ```Ruby
> 
>   def json_safe_and_pretty(hash_or_string)
>     hash_value = hash_or_string.is_a?(String) ? JSON.parse(hash_or_string) : hash_or_string
> ```

djudji [1:57 AM] 
OK, so this is the thing, the code brakes at AppointmentsList component, after the update. It gives an _appointments.map is not a function_ error, which means that it can not iterate over the Object type (edited)

[1:59]  
@justin any recommendations?

djudji [3:31 AM] 
could it be a new way of passing data from Rails to React (as I am testing more and more, it looks like props is not getting the right values from a controller)?

djudji [3:53 AM] 
it hangs inside function render(el, railsContext) in clientStartup.js

[3:55]  
this is how I am sending the data to a React component
```react_component 'Appointments', props: {appointments: @appointments}```

[3:56]  
and on "the other side" I get
```Object {appointments: "#<Appointment::ActiveRecord_Relation:0x007f972440cb90>"}```
instead of an array of Objects, like before updating to 6.9.1 (edited)

justin [8:16 AM] 
Dev or production. Please try both.

djudji [9:49 AM] 
Currently on dev. Will try it in production again (first app didn't work because I haven't set a new version for react_on_rails in package.json)

OK, this is a bit strange.

1. I cloned non-updated app
2. I did upgrade to react_on_rails 6.9.1
3. Created a new heroku app and pushed the upgraded app to Heroku.

Now everything is working fine in production, but in dev I still have the same error.

This is the URL -> https://calendar-react-on-rails.herokuapp.com/

I can confirm that both of the apps work when pushed to production.

This is the first URL -> https://calendar-railsreact.herokuapp.com/ (edited


the error is in the conversion to the pretty JSON… for development mode.

 ```  def json_safe_and_pretty(hash_or_string)
    if Rails.env.development?
      # TODO: for json_safe_and_pretty
      # 1. Add test
      # 2. Add error handler if cannot parse the string with nice message
      # 3. Consider checking that if not a string then a Hash
      hash_value = hash_or_string.is_a?(String) ? JSON.parse(hash_or_string) : hash_or_string
      ERB::Util.json_escape(JSON.pretty_generate(hash_value))
    else
      ERB::Util.json_escape(hash_or_string.to_json)
    end
  end
```


Before, we used to call:

```  def sanitized_props_string(props)
    props.is_a?(String) ? json_escape(props) : props.to_json
  end```


So I think if a Hash is passed, rather than a string, and we want to pretty print, then we’d have to go:

1. Hash => to_json (converts active record objects) => String
2. String => Hash
3. Then call `ERB::Util.json_escape(JSON.pretty_generate(hash_value))`

This could be a little slow.

1. Wondering if this should be optional, in `config/initializers/react_on_rails.rb`
2. Another option is skip pretty printing completely (but then should make an option).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/shakacode/react_on_rails/787#-:-KggFA9_7z0FTL11MGXB:b-u1x6lu)*
<!-- Sent from Reviewable.io -->

justin808 commented:
Note #789 is removing the pretty formatting.

We can add it back in...

justin808 commented:
@cheremukhin23 I thought more about the pretty formatting of json.

Let's add a configuration flag for this, since it tends to break...

Default is false.

```ruby
# Pretty printing props looks nice when you view the page source. However, if you have a large
# amount of props data, the page rendering can be a little slower.
config.pretty_print_props_in_development = false
```

justin808 commented:
See #791. I fixed another critical bug...

```
[1] (pry) #<#<Class:0x007f84e5bd19f8>>: 0> puts hash_or_string.to_json
"{\"comments\":[{\"id\":9,\"author\":\"xxx\",\"text\":\"xxxxxx\",\"created_at\":\"2017-01-11T08:14:10.296Z\",\"updated_at\":\"2017-01-11T08:14:10.296Z\"},

[2] (pry) #<Class:0x007f84e5bd19f8>>: 0> puts hash_or_string
{"comments":[{"id":9,"author":"xxx","text":"xxxxxx","created_at":"2017-01-11T08:14:10.296Z","updated_at":"2017-01-11T08:14:10.296Z"},
```

justin808 commented:
See https://github.com/shakacode/react_on_rails/issues/799

cheremukhin23 commented:
Review status: 0 of 2 files reviewed at latest revision, 5 unresolved discussions.

---

*[app/helpers/react_on_rails_helper.rb, line 223 at r2](https://reviewable.io:443/reviews/shakacode/react_on_rails/787#-KgchU4ZQq1Xoe8SFvfI:-KhSNPMD4Wb7C5iVnPZm:b-9jwz5i) ([raw file](https://github.com/shakacode/react_on_rails/blob/a3ef532d26039dc87d3032142a3171a6d26b2f3a/app/helpers/react_on_rails_helper.rb#L223)):*
<details><summary><i>Previously, justin808 (Justin Gordon) wrote…</i></summary><blockquote>

Can we throw here if the value passed is not a hash or a string? What else could the value be?
</blockquote></details>

We discard this code

---

*[app/helpers/react_on_rails_helper.rb, line 224 at r2](https://reviewable.io:443/reviews/shakacode/react_on_rails/787#-KggEw9tKZ5IIRJ7FLl0:-KhSNZQbe8Ghcccrwqf2:bszcpn5) ([raw file](https://github.com/shakacode/react_on_rails/blob/a3ef532d26039dc87d3032142a3171a6d26b2f3a/app/helpers/react_on_rails_helper.rb#L224)):*
<details><summary><i>Previously, justin808 (Justin Gordon) wrote…</i></summary><blockquote>

djudji [1:57 AM] 
OK, so this is the thing, the code brakes at AppointmentsList component, after the update. It gives an _appointments.map is not a function_ error, which means that it can not iterate over the Object type (edited)

[1:59]  
@justin any recommendations?

djudji [3:31 AM] 
could it be a new way of passing data from Rails to React (as I am testing more and more, it looks like props is not getting the right values from a controller)?

djudji [3:53 AM] 
it hangs inside function render(el, railsContext) in clientStartup.js

[3:55]  
this is how I am sending the data to a React component
```react_component 'Appointments', props: {appointments: @appointments}```

[3:56]  
and on "the other side" I get
```Object {appointments: "#<Appointment::ActiveRecord_Relation:0x007f972440cb90>"}```
instead of an array of Objects, like before updating to 6.9.1 (edited)

justin [8:16 AM] 
Dev or production. Please try both.

djudji [9:49 AM] 
Currently on dev. Will try it in production again (first app didn't work because I haven't set a new version for react_on_rails in package.json)

OK, this is a bit strange.

1. I cloned non-updated app
2. I did upgrade to react_on_rails 6.9.1
3. Created a new heroku app and pushed the upgraded app to Heroku.

Now everything is working fine in production, but in dev I still have the same error.

This is the URL -> https://calendar-react-on-rails.herokuapp.com/

I can confirm that both of the apps work when pushed to production.

This is the first URL -> https://calendar-railsreact.herokuapp.com/ (edited


the error is in the conversion to the pretty JSON… for development mode.

 ```  def json_safe_and_pretty(hash_or_string)
    if Rails.env.development?
      # TODO: for json_safe_and_pretty
      # 1. Add test
      # 2. Add error handler if cannot parse the string with nice message
      # 3. Consider checking that if not a string then a Hash
      hash_value = hash_or_string.is_a?(String) ? JSON.parse(hash_or_string) : hash_or_string
      ERB::Util.json_escape(JSON.pretty_generate(hash_value))
    else
      ERB::Util.json_escape(hash_or_string.to_json)
    end
  end
```


Before, we used to call:

```  def sanitized_props_string(props)
    props.is_a?(String) ? json_escape(props) : props.to_json
  end```


So I think if a Hash is passed, rather than a string, and we want to pretty print, then we’d have to go:

1. Hash => to_json (converts active record objects) => String
2. String => Hash
3. Then call `ERB::Util.json_escape(JSON.pretty_generate(hash_value))`

This could be a little slow.

1. Wondering if this should be optional, in `config/initializers/react_on_rails.rb`
2. Another option is skip pretty printing completely (but then should make an option).
</blockquote></details>

currently decided not to use pretty generate

---

*[app/helpers/react_on_rails_helper.rb, line 231 at r2](https://reviewable.io:443/reviews/shakacode/react_on_rails/787#-KgcggrDdvDqTLt3uR8q:-KhSNfG5p3IgHg_wp0bx:b-vjole3) ([raw file](https://github.com/shakacode/react_on_rails/blob/a3ef532d26039dc87d3032142a3171a6d26b2f3a/app/helpers/react_on_rails_helper.rb#L231)):*
<details><summary><i>Previously, justin808 (Justin Gordon) wrote…</i></summary><blockquote>

Can we move this catch block around the JSON.parse above?

we can do something like:

```ruby
  hash_value = if hash_or_string.is_a?(String) 
                          begin
                              JSON.parse(hash_or_string)
                           rescue JSON::ParserError => e
                               raise e, %(Cannot parse #{hash_or_string} \n Backtrace: #{e.backtrace.join("\n")})
                          else
                              hash_or_string
                          end  

```


</blockquote></details>

discarded code

---

*[app/helpers/react_on_rails_helper.rb, line 242 at r2](https://reviewable.io:443/reviews/shakacode/react_on_rails/787#-KgcgcN4lVDUvQARMTnr:-KhSNk5wKsnRHqy5t696:bj62ggv) ([raw file](https://github.com/shakacode/react_on_rails/blob/a3ef532d26039dc87d3032142a3171a6d26b2f3a/app/helpers/react_on_rails_helper.rb#L242)):*
<details><summary><i>Previously, justin808 (Justin Gordon) wrote…</i></summary><blockquote>

Can you provide a comment with a link on to the github source of this?
</blockquote></details>

added link

---

*[spec/dummy/spec/helpers/react_on_rails_helper_spec.rb, line 40 at r2](https://reviewable.io:443/reviews/shakacode/react_on_rails/787#-KgchiN8u2axRSMV9z4x:-KhSNnoJ0keOVuxHuWk4:b-qcgqp6) ([raw file](https://github.com/shakacode/react_on_rails/blob/a3ef532d26039dc87d3032142a3171a6d26b2f3a/spec/dummy/spec/helpers/react_on_rails_helper_spec.rb#L40)):*
<details><summary><i>Previously, justin808 (Justin Gordon) wrote…</i></summary><blockquote>

Can we add a couple tests that confirm the exceptions:

1. non-hash, non-string passed
2. non parseable string in development mode


Also, we're never testing the part where we check we're in development mode.

I'd try to mock/stub out the Rails.env for that test.

```ruby
# RSpec version
it 'tests my code' do
  allow(Rails).to receive(:env).and_return(ActiveSupport::StringInquirer.new('development'))
  # test code
end
```

[Some options from Google](https://www.google.com/search?q=stub+rails+env+for+test+to+production)


</blockquote></details>

added test for arguments

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/shakacode/react_on_rails/787)*
<!-- Sent from Reviewable.io -->

justin808 commented:
We need to fix this deprecation for React.

```
====> React On Rails: Checking app/assets/webpack for outdated/missing bundles
You're running an old version of PhantomJS, update to >= 2.1.1 for a better experience.
  the hello world example works (FAILED - 1)
Failures:
  1) Hello World the hello world example works
     Failure/Error: visit "/hello_world"
     
     Capybara::Poltergeist::JavascriptError:
       One or more errors were raised in the Javascript code on the page. If you don't care about these errors, you can ignore them by setting js_errors: false in your Poltergeist configuration (see documentation for details).
     
       Warning: Accessing PropTypes via the main React package is deprecated. Use the prop-types package from npm instead.
       Warning: Accessing PropTypes via the main React package is deprecated. Use the prop-types package from npm instead.
           at :36 in printWarning
     # /home/travis/.rvm/gems/ruby-2.2.6/gems/poltergeist-1.13.0/lib/capybara/poltergeist/browser.rb:376:in `command'
     # /home/travis/.rvm/gems/ruby-2.2.6/gems/poltergeist-1.13.0/lib/capybara/poltergeist/browser.rb:35:in `visit'
     # /home/travis/.rvm/gems/ruby-2.2.6/gems/poltergeist-1.13.0/lib/capybara/poltergeist/driver.rb:97:in `visit'
     # /home/travis/.rvm/gems/ruby-2.2.6/gems/capybara-2.12.1/lib/capybara/session.rb:252:in `visit'
     # /home/travis/.rvm/gems/ruby-2.2.6/gems/capybara-2.12.1/lib/capybara/dsl.rb:52:in `block (2 levels) in <module:DSL>'
     # ./spec/features/hello_world_spec.rb:5:in `block (2 levels) in <top (required)>'
     # /home/travis/.rvm/gems/ruby-2.2.6/gems/rspec-retry-0.5.3/lib/rspec/retry.rb:112:in `block in run'
     # /home/travis/.rvm/gems/ruby-2.2.6/gems/rspec-retry-0.5.3/lib/rspec/retry.rb:101:in `loop'
     # /home/travis/.rvm/gems/ruby-2.2.6/gems/rspec-retry-0.5.3/lib/rspec/retry.rb:101:in `run'
     # /home/travis/.rvm/gems/ruby-2.2.6/gems/rspec-retry-0.5.3/lib/rspec_ext/rspec_ext.rb:12:in `run_with_retry'
     # /home/travis/.rvm/gems/ruby-2.2.6/gems/rspec-retry-0.5.3/lib/rspec/retry.rb:30:in `block (2 levels) in setup'
Finished in 2.4 seconds (files took 2.58 seconds to load)
1 example, 1 failure
```

justin808 commented:
Looking good:

1. Few comments.
2. Need to pass CI. That could be a separate PR on master
3. Need a CHANGELOG.md entry.

---

Reviewed 2 of 2 files at r3.
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks broke.

---

*[app/helpers/react_on_rails_helper.rb, line 234 at r3](https://reviewable.io:443/reviews/shakacode/react_on_rails/787#-KhTLldRTCKIUj_MBtLG:-KhTLldRTCKIUj_MBtLH:b-xehhej) ([raw file](https://github.com/shakacode/react_on_rails/blob/358d5726a8d6a81dede8c939de7ee3a7248d9d1d/app/helpers/react_on_rails_helper.rb#L234)):*
> ```Ruby
> 
>   def escape_json(json)
>     return old_json_escape(json) unless Rails::VERSION::STRING >= "4"
> ```

I think this not right. What if the Rails version is "10.0".

Please google for a proper version comparator.

Something like this (you creating the `rails_version_less_than`

```
 return old_json_escape(json) if rails_version_less_than(4.0)
```

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/shakacode/react_on_rails/787#-:-KhTMlxi7KrJ8FjAdIfV:bysbhdf)*
<!-- Sent from Reviewable.io -->

justin808 commented:
@cheremukhin23 See comment: https://github.com/shakacode/react_on_rails/issues/799#issuecomment-293706951

justin808 commented:
@cheremukhin23 @squadette Maybe I should merge this one and let's add the feature in a new PR?

justin808 commented:
please rebase on master and that should fix the CI, @cheremukhin23

cheremukhin23 commented:
Review status: 1 of 3 files reviewed at latest revision, 1 unresolved discussion.

---

*[app/helpers/react_on_rails_helper.rb, line 234 at r3](https://reviewable.io:443/reviews/shakacode/react_on_rails/787#-KhTLldRTCKIUj_MBtLG:-Khc2bW_J9ESAjOJarlz:b-6sqq66) ([raw file](https://github.com/shakacode/react_on_rails/blob/358d5726a8d6a81dede8c939de7ee3a7248d9d1d/app/helpers/react_on_rails_helper.rb#L234)):*
<details><summary><i>Previously, justin808 (Justin Gordon) wrote…</i></summary><blockquote>

I think this not right. What if the Rails version is "10.0".

Please google for a proper version comparator.

Something like this (you creating the `rails_version_less_than`

```
 return old_json_escape(json) if rails_version_less_than(4.0)
```
</blockquote></details>

Used Gem::Version to compare versions

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/shakacode/react_on_rails/787)*
<!-- Sent from Reviewable.io -->

justin808 commented:
Very close. But missing tests on part of the PR.

---

Reviewed 1 of 1 files at r5, 1 of 1 files at r6.
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[app/helpers/react_on_rails_helper.rb, line 251 at r6](https://reviewable.io:443/reviews/shakacode/react_on_rails/787#-KhccYzOx23mj82n3o29:-KhccYzOx23mj82n3o2A:bih7izj) ([raw file](https://github.com/shakacode/react_on_rails/blob/19722db3260512c8ddd6e3630f2b20975534bd62/app/helpers/react_on_rails_helper.rb#L251)):*
> ```Ruby
>     json_escape = { "&" => '\u0026', ">" => '\u003e', "<" => '\u003c', "\u2028" => '\u2028', "\u2029" => '\u2029' }
>     json_escape_regexp = /[\u2028\u2029&><]/u
>     json.to_s.gsub(json_escape_regexp, json_escape)
> ```

This code is not tested. We need a test to confirm this method works as well as the method rails_version_less_than.

Please add a couple simple tests for rails_version_less_than with cases "3.2", "3", "4.2", "10.0".

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/shakacode/react_on_rails/787#-:-Khccw3ENOYG7S3aCG5e:b27l0v1)*
<!-- Sent from Reviewable.io -->

justin808 commented:
@cheremukhin23 Please let me know if you can finish this. If you don't have time, I need to find an alternate resource. 

@squadette, any chance that you can help with this one?

justin808 commented:
@ynote, is there chance that you can give us a hand with this one?

djfdev commented on app/helpers/react_on_rails_helper.rb:
```diff
@@ -224,8 +224,33 @@ def server_render_js(js_expression, options = {})
 
+  def json_safe_and_pretty(hash_or_string)
+    unless hash_or_string.class.in?([Hash, String])
+      raise "#{hash_or_string.class} is unsupported argument class for this method"
```
I would name the types that are accepted by the method in this error.

Ynote commented:
Hi,

I can help on the PR. @cheremukhin23, can you give me permission on the PR so I can fetch the branch and add the missing tests?

justin808 commented:
@Ynote @cheremukhin23 You can also pull his changes into your own fork and make a new branch.

@squadette would like us to put back in JSON props as pretty. I'm ok with that either in this or a different PR, but I won't accept any PR without full test coverage and documentation changes.

justin808 commented:
See #812.
